src/data/finsupp/convolution.lean

```diff
@@ -0,0 +1,266 @@
+by simp only [mul_def, sum_zero_index]
+
+private lemma mul_zero (f : monoid_algebra α β) : f * 0 = 0 :=
+by simp only [mul_def, sum_zero_index, sum_zero]
+
+private lemma left_distrib (a b c : monoid_algebra α β) : a * (b + c) = a * b + a * c :=
+by simp only [mul_def, sum_add_index, mul_add, _root_.mul_zero, single_zero, single_add,
+  eq_self_iff_true, forall_true_iff, forall_3_true_iff, sum_add]
+
+private lemma right_distrib (a b c : monoid_algebra α β) : (a + b) * c = a * c + b * c :=
+by simp only [mul_def, sum_add_index, add_mul, _root_.mul_zero, _root_.zero_mul, single_zero,
+  single_add, eq_self_iff_true, forall_true_iff, forall_3_true_iff, sum_zero, sum_add]
+
+instance : semiring (monoid_algebra α β) :=
+{ one       := 1,
+  mul       := (*),
+  one_mul   := assume f, by simp only [mul_def, one_def, sum_single_index, _root_.zero_mul,
+    single_zero, sum_zero, zero_add, one_mul, sum_single],
+  mul_one   := assume f, by simp only [mul_def, one_def, sum_single_index, _root_.mul_zero,
+    single_zero, sum_zero, add_zero, mul_one, sum_single],
+  zero_mul  := zero_mul,
+  mul_zero  := mul_zero,
+  mul_assoc := assume f g h, by simp only [mul_def, sum_sum_index, sum_zero_index, sum_add_index,
+    sum_single_index, single_zero, single_add, eq_self_iff_true, forall_true_iff, forall_3_true_iff,
+    add_mul, mul_add, add_assoc, mul_assoc, _root_.zero_mul, _root_.mul_zero, sum_zero, sum_add],
+  left_distrib  := left_distrib,
+  right_distrib := right_distrib,
+  .. finsupp.add_comm_monoid }
+
+end
+
+instance [comm_monoid α] [comm_semiring β] : comm_semiring (monoid_algebra α β) :=
+{ mul_comm := assume f g,
+  begin
+    simp only [mul_def, finsupp.sum, mul_comm],
+    rw [finset.sum_comm],
+    simp only [mul_comm]
+  end,
+  .. monoid_algebra.semiring }
+
+instance [ring β] : has_neg (monoid_algebra α β) :=
+by apply_instance
+
+instance [monoid α] [ring β] : ring (monoid_algebra α β) :=
+{ neg := has_neg.neg,
+  add_left_neg := add_left_neg,
+  .. monoid_algebra.semiring }
+
+instance [comm_monoid α] [comm_ring β] : comm_ring (monoid_algebra α β) :=
+{ mul_comm := mul_comm, .. monoid_algebra.ring}
+
+lemma single_mul_single [monoid α] [semiring β] {a₁ a₂ : α} {b₁ b₂ : β}:
+  (single a₁ b₁ : monoid_algebra α β) * single a₂ b₂ = single (a₁ * a₂) (b₁ * b₂) :=
+(sum_single_index (by simp only [_root_.zero_mul, single_zero, sum_zero])).trans
+  (sum_single_index (by rw [_root_.mul_zero, single_zero]))
+
+universe ui
+variable {ι : Type ui}
+
+lemma prod_single [comm_monoid α] [comm_semiring β]
+  {s : finset ι} {a : ι → α} {b : ι → β} :
+  s.prod (λi, single (a i) (b i)) = single (s.prod a) (s.prod b) :=
+finset.induction_on s rfl $ λ a s has ih, by rw [prod_insert has, ih,
+  single_mul_single, prod_insert has, prod_insert has]
+
+end monoid_algebra
+
+section
+variables [semiring β]
+
+/-- a type synonym for `finsupp α β`, on which we define a multiplication by convolution -/
+@[derive inhabited]
+def add_monoid_algebra := α →₀ β
```

Comment:
Could this be defined as `monoid_algebra (multiplicative alpha) beta`? That would probably make a lot of type classes into one liners.

Reply:
No. I tried this, and it was an unholy mess in the `polynomials.lean` file. Depending on inscrutable details of what Lean was thinking at the time, sometimes it was "seeing through" `multiplicative nat = nat`, and sometimes it wasn't.

If we want to do this, we would have to bite the bullet and replace `multiplicative` with an inductive type, and explicitly use `multiplicative.wrap` and `multiplicative.unwrap` whenever we want to cross the barrier. This may be a good idea in any case ...

Reply:
Hmmm, it's unfortunate that we can't do
`local attribute [hyperdupermegeirreducible] multiplicative`